torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp

```diff
@@ -0,0 +1,578 @@
+      // producers
+      std::unordered_set<IterDomain*> mapped_c_ids_left_of_ca;
+
+      auto tv_inputs = ir_utils::filterByType<TensorView>(expr->inputs());
+
+      for (auto p_tv : tv_inputs) {
+        // If outside computeAt axis, we don't want to directly map
+        // consumer/producer as their thread mappings could change as long as
+        // it's across shared/global memory.
+
+        // Mark axes outside compute at point for parallel type tracking
+        std::unordered_set<IterDomain*> right_of_ca_point;
+        if (mapping_mode_ == MappingMode::PARALLEL &&
+            p_tv->getThisComputeAtAxis() < p_tv->nDims()) {
+          right_of_ca_point.insert(
+              p_tv->domain()->domain().begin() + p_tv->getThisComputeAtAxis(),
+              p_tv->domain()->domain().end());
+        }
+        // if this is a producer tv, (i.e. not a terminating output tv), then
+        // produce at is the same as this compute at position. Loop mode does
+        // its own thing, see below in this function.
+        if (mapping_mode_ != MappingMode::LOOP) {
+          produce_at_map_[p_tv] = p_tv->getThisComputeAtAxis();
+        }
+
+        auto c2p_root_map =
+            PairwiseRootDomainMap(p_tv, c_tv)
+                .mapConsumerToProducer(c_tv->domain(), p_tv->domain());
+
+        // Look for matching ID transformations in producer and consumer, replay
+        // producer as consumer. We want to play producer as consumer instead of
+        // the other way around since consumer may have some broadcasted axes
+        // producer doesn't have merged into loops producer may use. If we did
+        // consumer as producer we wouldn't have this information in the
+        // mapping. If we're using this map for indexing, we do not want to
+        // propagate broadcast mismatches. If we're using it to identify loop
+        // nests, we do want to propagate mismatches.
+        BestEffortReplay replay_PasC(
+            p_tv->domain()->domain(),
+            c_tv->domain()->domain(),
+            c2p_root_map,
+            mapping_mode_ == MappingMode::LOOP);
+
+        auto c2p_map = replay_PasC.getReplay();
+
+        // Map the entire replay map
+        // Also reverse the map, as we use p2c_map to find this computeAt
+        // position in consumer. This could be removed if we changed computeAt
+        // of TensorViews
+        std::unordered_map<IterDomain*, IterDomain*> p2c_map;
+        for (auto entry : c2p_map) {
+          auto c_id = entry.first;
+          auto p_id = entry.second;
+          // If outside CA point and we're creating parallel map, do not map the
+          // axis
+          if (mapping_mode_ == MappingMode::PARALLEL &&
+              right_of_ca_point.find(p_id) != right_of_ca_point.end()) {
+            continue;
+          }
+          // Map the id's together
+          map_ids(p_id, c_id);
+          p2c_map[p_id] = c_id;
```

Comment:
c2p_map is not one-to-one when the mapping mode is LOOP, i.e., when bcast mismatch is allowed. Doesn't it cause p2c_map non-deterministic? There can be multiple c_id for a single p_id. The final mapping for the p_id would depend on the internal ordering of c2p_map.

Reply:
There are multiple p_id's it could map to, but the mapping will simply forward from best effort replay. So the most bcast merged axis in producer should be mapped to the corresponding c_id.
I'm not 100%, but I'm pretty sure it should be deterministic.

Reply:
Apparently, there are multiple different c_ids for a single p_id. I inserted this:

```
diff --git a/torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp b/torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp
index 6001e02dc5..c320ee9781 100644
--- a/torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp
+++ b/torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp
@@ -287,6 +287,7 @@ void ComputeAtMap::build() {
           }
           // Map the id's together
           mapIds(p_id, c_id);
+          TORCH_INTERNAL_ASSERT(p2c_map.find(p_id) == p2c_map.end() || p2c_map.at(p_id) == c_id);
           p2c_map[p_id] = c_id;
         }
```

And I got assertion failures with it. So, the mapping of the p_id would just depend on the ordering among the multiple c_ids, which would be non-deterministic.

Reply:
Fixed this with https://github.com/csarofeen/pytorch/pull/597/commits/6f279a2ecb5a135b4ef9d99fbfacb75af60c8673